sdig.8
.TH SDIG 8 "Mon Apr  4 2003" "" "Switch Digger"
.SH NAME
sdig \- The Switch Digger
.SH SYNOPSIS
.B sdig [-d] [-v] [-F] [-p/-P] [-f \fIconfig\fB] [-m/-m \fIMAC\fB] ( \fIIP\fB | \fIhostname\fB )

.SH DESCRIPTION

The Switch Digger, or sdig, is a tool that is intended to help network
administrators track down systems.  It was designed in a public school
district environment with about 1500 systems spread across 25 remote
locations.

sdig works by first finding the IP address of the target system, then it
contacts the router(s) in that network to get the MAC address for that IP
address.  With that known, it then probes every switch on the target
network to find a port number.  The port that doesn't lead to another
switch is returned, along with any description you may have provided.

.SH OPTIONS

.IP "-d"
Raise the debugging level by 1.  This gets rather messy above 3 or 4.

.IP "-v"
Be verbose.  This makes sdig print every port instead of just the one that
is the most likely candidate, for example (includes inter-switch ports 
with LINKINFO written in sdig.conf).

.IP "-F"
Fast mode.  sdig will not do reverse DNS (in-addr.arpa) or NetBIOS queries
to port 137 when this is enabled.

.IP "-f \fIconfig\fR"
Use the configuration file \fIconfig\fR.

.IP "-m \fIMAC\fR"
Look for this MAC address rather than asking a router about it.  You still
can provide an IP address or hostname so that sdig knows which network
to check.

.IP "-m"
A total-network sweep option is when you don't provide the IP, takes longer
to query all switches, so care is taken than each IPxCOMMUNITY is only
queried once.

.IP \fIIP\fR
An Internet Protocol address to find, i.e. 192.168.1.1.

.IP \fIhostname\fR
A DNS or WINS hostname to find.  WINS resolution is only available if you
have installed nmblookup from Samba and have added it to your config file.

.IP "-p/-P"
Parallelized SNMP queries have been added and improved as a feature of
the recent sdig versions. If compiled in, they can be disabled at run time,
or different activities may be done at discretion of future programmers.
"-p" increases the "use parallelism" counter.
"-P" decreases the "use parallelism" counter, but to no less than zero.
You might want to disable this i.e. if it misbehaves on your platform,
or if you have very many switches and spawning many children would exhaust
your file descriptors (network sockets) or process table entries.
Hint: Future versions may add a limit on number of spawned children.
If the "parallelized queries" feature is not compiled in, these "-p/-P"
flags are recognized, but ignored.

.SH LIMITATIONS
You can't track down arbitrary hosts on the Internet.  Well, most people
can't.  You might be able to do this if you convince all the router and
switch manufacturers of the world to drop in a SNMP backdoor for your
sdig host.  US government three letter entities: contact me for details.

.SH BUGS
This program was developed on just one kind of system (Linux glibc2) 
so it probably doesn't compile cleanly on others.

NOTE: version 0.45 was developed on Solaris x86/SPARC and also works there.

.SH BACKGROUND
I (Russell Kroll) first wrote this program to show some local people 
that you don't need to dump lots of money into a program like 3com's 
Transcend just to hunt down some lusers on your network.  If you don't 
need to create fancy network diagrams to impress the PHBs, then this 
program will probably work for you.

It was developed originally on 3com SuperStack 3300s, and continues
to be tested both on those and various HP 2324s and 4108s.  Other
equipment should also work if it provides the same basic OIDs.

Jim Klimov also tested it in a diverse network with HP, Cisco, Avaya,
and Allied Telesyn equipment, to name a few.

.SH SEE ALSO
\fBsdig.conf\fR(5)

.SH AUTHORS
Russell Kroll <rkroll@exploits.org>     up till sdig-0.40
Russell A. Jackson <raj@csub.edu>       sdig-0.41 .. sdig-0.44
Jim Klimov <jimklimov@gmail.com>        sdig-0.45


NEWS
Release notes for 0.45

Maintainers have changed, the project moved to SourceForge, and a host of
new features were added (see ChangeLog). In particular, you can now do
parallelized SNMP searches over several devices simultaneously, search
for a MAC address on all switches, and use different IPs (or hostnames)
for SNMP queries and as routing interfaces of ROUTER's.

Tested to work on Linux and Solaris.

Release notes for 0.40

The build system now uses autoconf, so you can do the usual tricks with 
--prefix and --sysconfdir to customize your build.  Be aware that the 
default configuration will look for sdig.conf in /usr/local/etc, while 
previous versions assumed /etc.

DESTDIR is now supported for packaging purposes.

SNMPv2-MIB::sysName.0 is now printed when available for switches.

nbname has been dropped.  If you still need additional NetBIOS type data, 
write a script to call Samba's nmblookup, then set it as your HOSTINFO 
command.

UCD-SNMP support is essentially gone.  You can probably still make it work 
if you're determined to use the old libraries, but it will require 
renaming a bunch of includes and variable types.

The -m switch to search by MAC addresses works again.


src/sdig.c
/* sdig.c - the Switch Digger main file
 * Current version is on SourceForge:   http://sdig.sourceforge.net/
 *
 *  Copyright (C) 2000-2003  Russell Kroll <rkroll@exploits.org>
 *	    up till sdig-0.40
 *  Copyright (C) 2005-2006  Russell Jackson <raj@csub.edu>
 *	    sdig-0.41 .. sdig-0.44
 *  Copyright (C) 2010  Jim Klimov <jimklimov@gmail.com>
 *	    sdig-0.45 .. 0.46 (in progress)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *  $Id$
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <ctype.h>
#include <errno.h>
#include <netdb.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h> 
#include <string.h> 
#include <unistd.h> 
#include <arpa/inet.h> 
#include <netinet/in.h> 
#include <sys/socket.h>
#include <sysexits.h>

#include "sdig.h"
#include "common.h"
#include "snmpget.h"

/*
 * Flags
 */

int	verbose = 0, fastmode = 0, dofork = 0;

void
help(const char *prog)
{
	printf("SNMP-based router and switch probe for locating client systems.\n\n");
	printf("usage: %s [-d] [-f <config>] [-m <MAC>] [-v] [-F] (<IP> | <hostname>)\n", prog);
	printf("\n");
	printf("  -d		- increase debug level\n");
	printf("  -F		- fast mode - no DNS/NetBIOS reverse lookups\n");	
	printf("  -f <config>	- use alternate config <config>\n");
	printf("                  default %s/sdig.conf\n", "CONFPATH");
	printf("  -m <MAC>	- force MAC <MAC>, xx:xx:xx:xx:xx:xx format\n");
	printf("  -v		- be verbose\n");
	printf("  <IP>		- IP address to find\n");
	printf("  <hostname>	- DNS/WINS hostname to find\n");
	printf("  -p / -P	- enable / disable forking children for SNMP queries\n");
#ifdef SDIG_USE_SEMS
	printf("  		  (NOTE: enabled by default)\n");
#else
	printf("  		  (NOTE: feature NOT COMPILED into this binary)\n");
#endif

	exit(EX_OK);
}

int
main(int argc, char *argv[])
{
	char	*prog, *query, *conf = NULL, *mac = NULL, *stdmac = NULL;
	int	i;

	printf("Switch Digger %s", VERSION);
#ifdef SDIG_USE_SEMS
	printf(", query forking capable", VERSION);
	dofork=1;
#else
	dofork=0;
#endif
	printf("\n\n");

	prog = argv[0];

	while ((i = getopt(argc, argv, "+dpPhf:m:vF")) != EOF) {
		switch (i) {
			case 'd':
				inc_debuglevel();
				break;

			case 'p':
#ifdef SDIG_USE_SEMS
				dofork++;
#else
				printf("ERROR: query forking not compiled in, '-p' ignored\n");
#endif
				break;

			case 'P':
#ifdef SDIG_USE_SEMS
				dofork--;
#else
				printf("ERROR: query forking not compiled in, '-P' ignored\n");
#endif
				break;

			case 'f':
				conf = optarg;
				break;

			case 'h':
				help(prog);
				break;

			case 'm':
				mac = optarg;
				break;

			case 'v':
				verbose++;
				break;

			case 'F':
				fastmode = 1;
				break;
				
			default:
				help(prog);
		}
	}

	argc -= optind;
	argv += optind;

	if (argc < 1 && !(mac))
		help(prog);

	if (argc == 0) {
		query = NULL;
	} else {
	query = argv[0];
	}

	loadconfig(conf);
#ifdef SDIG_USE_SEMS
	if ( dofork < 0 ) { dofork = 0; }
	printf("Query forking is currently %s\n\n", (dofork?"enabled":"disabled") );
	if ( dofork ) { output_sem_init(); }
#endif

	/* split off to resolve things based on what kind of input we got */

	/* hostname (DNS or WINS) given */
	if ((query) && !isip(query)) {
		printf("    Query: %s\n", query);
		resolvename(query);

		/* NOTREACHED */
	}

	/* MAC address specified, along with target network */
	if ((mac)) {
		stdmac = standardize_mac(mac);

		if ((query) && (isip(query))) {
			printf("    Query: MAC %s in network %s\n", 
				stdmac, query);

			switchscan(query, pack_mac(stdmac));

		/* NOTREACHED */
		} else {
			/* Only a MAC is provided */
			printf("    Query: MAC %s in any network\n", 
				stdmac);

			switchscan(NULL, pack_mac(stdmac));

			/* NOTREACHED */
		}
	}

	/* just an IP address given */
	if (isip(query)) {
		printf("    Query: %s\n", query);
		routerscan(query);

		/* NOTREACHED */
	}

	/* unknown! */
	fprintf(stderr, "Error: unknown query type!\n");
	exit(1);
}
